File: app/src/main/java/com/github/gustavofernandes/radioplayer/AudioService.java
package com.github.gustavofernandes.radioplayer;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.NotificationCompat;
import android.support.v4.media.MediaBrowserCompat;
import android.support.v4.media.MediaBrowserServiceCompat;
import android.support.v4.media.session.MediaButtonReceiver;
import android.support.v4.media.session.MediaSessionCompat;
import android.support.v4.media.session.PlaybackStateCompat;

import java.util.Collections;
import java.util.List;

public class AudioService extends MediaBrowserServiceCompat {

    public static final String ROOT = "root";

    private static final String TAG = AudioService.class.getSimpleName();

    private MediaSessionCompat mMediaSession;
    private PlaybackStateCompat.Builder mPlaybackStateBuilder;

    private MediaPlayer mMediaPlayer;

    private AudioManager mAudioManager;
    private AudioManagerFocusChangeListener mAudioManagerFocusChangeListener;

    private BroadcastReceiver mNoisyReceiver;

    private NotificationCompat.Builder mNotificationBuilder;

    @Override
    public void onCreate() {
        super.onCreate();

        mMediaSession = new MediaSessionCompat(this, TAG);
        mMediaSession.setFlags(
                MediaSessionCompat.FLAG_HANDLES_MEDIA_BUTTONS |
                        MediaSessionCompat.FLAG_HANDLES_TRANSPORT_CONTROLS);
        mMediaSession.setCallback(new MediaSessionCallback());

        setSessionToken(mMediaSession.getSessionToken());

        mPlaybackStateBuilder = new PlaybackStateCompat.Builder();
        mPlaybackStateBuilder
                .setState(PlaybackStateCompat.STATE_NONE, 0, 0)
                .setActions(PlaybackStateCompat.ACTION_PREPARE)
                .build();

        mMediaSession.setPlaybackState(mPlaybackStateBuilder.build());

        mMediaPlayer = new MediaPlayer();

        // TODO: move AudioManager and listener to own class?
        mAudioManager = (AudioManager) getSystemService(Context.AUDIO_SERVICE);
        mAudioManagerFocusChangeListener = new AudioManagerFocusChangeListener();

        mNoisyReceiver = new BroadcastReceiver() {

            @Override
            public void onReceive(Context context, Intent intent) {
                // TODO: pause playback
            }
        };
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        MediaButtonReceiver.handleIntent(mMediaSession, intent);
        return super.onStartCommand(intent, flags, startId);
    }

    @Override
    public void onDestroy() {
        mMediaPlayer.release();
        mMediaPlayer = null;
        mMediaSession.release();
    }

    @Override
    public BrowserRoot onGetRoot(@NonNull String clientPackageName, int clientUid, Bundle rootHints) {
        return new BrowserRoot(ROOT, null);
    }

    @Override
    public void onLoadChildren(@NonNull String parentId, @NonNull Result<List<MediaBrowserCompat.MediaItem>> result) {
        result.sendResult(getMediaItemsById(parentId));
    }

    private List<MediaBrowserCompat.MediaItem> getMediaItemsById(String id) {
        // TODO

        return Collections.emptyList();
    }

    private class MediaSessionCallback extends MediaSessionCompat.Callback {

        @Override
        public void onPlay() {
            int result = mAudioManager.requestAudioFocus(mAudioManagerFocusChangeListener, AudioManager.STREAM_MUSIC, AudioManager.AUDIOFOCUS_GAIN);
            if (result == AudioManager.AUDIOFOCUS_REQUEST_GRANTED) {

                mMediaSession.setActive(true);

                registerReceiver(mNoisyReceiver, new IntentFilter(AudioManager.ACTION_AUDIO_BECOMING_NOISY));

                // TODO: proceed with playing

                mPlaybackStateBuilder
                        .setState(PlaybackStateCompat.STATE_PLAYING, 0, 1)
                        .setActions(PlaybackStateCompat.ACTION_PLAY_PAUSE)
                        .build();
                mMediaSession.setPlaybackState(mPlaybackStateBuilder.build());

                // TODO: set MediaMetadataCompat on mMediaSession

                buildNotification(); // TODO: optimize
                startForeground(1, mNotificationBuilder.build());
            }
        }

        @Override
        public void onPause() {
            mMediaPlayer.pause();

            mPlaybackStateBuilder
                    .setState(PlaybackStateCompat.STATE_PAUSED, 0, 0)
                    .setActions(PlaybackStateCompat.ACTION_PLAY_PAUSE)
                    .build();
            mMediaSession.setPlaybackState(mPlaybackStateBuilder.build());

            // TODO: update notification

            stopForeground(false);
        }

        @Override
        public void onStop() {
            mAudioManager.abandonAudioFocus(mAudioManagerFocusChangeListener);
            mMediaSession.setActive(false);
            unregisterReceiver(mNoisyReceiver);

            stopForeground(true);
        }
    }

    private void buildNotification() {

        mNotificationBuilder = MediaSessionNotificationBuilder.Companion.from(this, mMediaSession);

        mNotificationBuilder.setStyle(new android.support.v4.media.app.NotificationCompat.MediaStyle()
            .setMediaSession(mMediaSession.getSessionToken())
            .setShowCancelButton(true)
            .setCancelButtonIntent(MediaButtonReceiver.buildMediaButtonPendingIntent(this, PlaybackStateCompat.ACTION_STOP)));
    }

    private class AudioManagerFocusChangeListener implements AudioManager.OnAudioFocusChangeListener {

        @Override
        public void onAudioFocusChange(int i) {
            if (i == AudioManager.AUDIOFOCUS_LOSS) {
                // TODO: implement this (stop playback here) and other changes
                // AUDIOFOCUS_LOSS_TRANSIENT, AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK, AUDIOFOCUS_GAIN
            }
        }
    }
}

File: app/src/main/java/com/github/gustavofernandes/radioplayer/AudioService.kt
package com.github.gustavofernandes.radioplayer

import android.content.Intent
import android.database.Cursor
import android.media.MediaPlayer
import android.os.Bundle
import android.provider.MediaStore
import android.support.v4.media.MediaBrowserCompat
import android.support.v4.media.MediaBrowserServiceCompat
import android.support.v4.media.MediaDescriptionCompat
import android.support.v4.media.session.MediaButtonReceiver
import android.util.Log
import java.util.*

class AudioService: MediaBrowserServiceCompat(), MediaSessionCallback {

    companion object {
        const val ROOT = "root"
        const val STATIONS = "stations"
    }

    private val TAG = AudioService::class.java.simpleName

    private lateinit var mediaSessionWrapper: MediaSessionWrapper

    private var mediaPlayer: MediaPlayer? = null

    override fun onCreate() {
        Log.d(TAG, "onCreate")
        super.onCreate()

        mediaSessionWrapper = MediaSessionWrapper(this, this)

        sessionToken = mediaSessionWrapper.sessionToken

        mediaPlayer = MediaPlayer()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        Log.d(TAG, "onStartCommand")
        MediaButtonReceiver.handleIntent(mediaSessionWrapper.mediaSession, intent)
        return super.onStartCommand(intent, flags, startId)
    }

    override fun onDestroy() {
        Log.d(TAG, "onDestroy")
        mediaSessionWrapper.onDestroy()
        mediaPlayer?.release()
        mediaPlayer = null
    }

    override fun onGetRoot(clientPackageName: String, clientUid: Int, rootHints: Bundle?): BrowserRoot? {
        Log.d(TAG, "onGetRoot")
        return BrowserRoot(ROOT, null)
    }

    override fun onLoadChildren(parentId: String, result: Result<List<MediaBrowserCompat.MediaItem>>) {
        Log.d(TAG, "onLoadChildren")

        val list = mutableListOf<MediaBrowserCompat.MediaItem>()

        if (parentId == STATIONS) {
            val cursor: Cursor? = contentResolver.query(
                    MediaStore.Audio.Media.EXTERNAL_CONTENT_URI,
                    null,
                    MediaStore.Audio.Media.DATA + " LIKE ?",
                    arrayOf("%/radioplayer-media/Stations/%"),
                    MediaStore.Audio.Media.DATA)

            when {
                cursor == null -> {
                    Log.d(TAG, "query failed")
                }
                !cursor.moveToFirst() -> {
                    Log.d(TAG, "no media found")
                }
                else -> {
                    var prevStationTitle: String? = null
                    do {
                        val absolutePath = cursor.getString(cursor.getColumnIndex(MediaStore.Audio.Media.DATA))
                        val stationIndex = absolutePath.indexOf("/radioplayer-media/Stations/") + "/radioplayer-media/Stations/".length
                        val stationTitle = absolutePath.substring(stationIndex, absolutePath.indexOf('/', stationIndex))

                        if (stationTitle != prevStationTitle) {

                            Log.d(TAG, "Found station titled: $stationTitle")

                            val mediaDescription = MediaDescriptionCompat.Builder()
                                    .setMediaId("$STATIONS - $stationTitle") // TODO: change mediaId to something more unique
                                    .setTitle(stationTitle)
                                    .build()

                            list.add(MediaBrowserCompat.MediaItem(mediaDescription, MediaBrowserCompat.MediaItem.FLAG_BROWSABLE))
                            prevStationTitle = stationTitle
                        }

                    } while (cursor.moveToNext())
                }
            }

            cursor?.close()
        }

        result.sendResult(list)
    }

    override fun onPlay() {
        Log.d(TAG, "onPlay")
        // TODO: play mediaplayer
        startForeground(1, mediaSessionWrapper.notificationBuilder.build())
    }

    override fun onPause() {
        Log.d(TAG, "onPause")
        // TODO: pause mediaplayer
        stopForeground(false)
    }

    override fun onStop() {
        Log.d(TAG, "onStop")
        stopForeground(true)
    }

    private fun getMediaItemsById(id: String): MutableList<MediaBrowserCompat.MediaItem>? {
        return Collections.emptyList() // TODO
    }
}
